app/src/main-process/main.ts

```diff
@@ -1,42 +1,95 @@
-import { app, Menu, MenuItem, ipcMain, BrowserWindow, autoUpdater } from 'electron'
+import { app, Menu, MenuItem, ipcMain, BrowserWindow, autoUpdater, dialog } from 'electron'
 
 import { AppWindow } from './app-window'
-import { buildDefaultMenu, MenuEvent, findMenuItemByID } from './menu'
+import { CrashWindow } from './crash-window'
+import { buildDefaultMenu, MenuEvent, findMenuItemByID, setCrashMenu } from './menu'
 import { parseURL } from '../lib/parse-url'
 import { handleSquirrelEvent } from './squirrel-updater'
 import { SharedProcess } from '../shared-process/shared-process'
 import { fatalError } from '../lib/fatal-error'
 
-import { showFallbackPage } from './error-page'
 import { IMenuItemState } from '../lib/menu-update'
 import { ILogEntry, logError, log } from '../lib/logging/main'
+import { formatError } from '../lib/logging/format-error'
+import { reportError } from './exception-reporting'
+import { enableSourceMaps } from '../lib/enable-source-maps'
+
+enableSourceMaps()
 
 let mainWindow: AppWindow | null = null
 let sharedProcess: SharedProcess | null = null
 
 const launchTime = Date.now()
 
+let preventQuit = false
 let readyTime: number | null = null
+let hasReportedUncaughtException = false
 
 type OnDidLoadFn = (window: AppWindow) => void
 /** See the `onDidLoad` function. */
 let onDidLoadFns: Array<OnDidLoadFn> | null = []
 
-process.on('uncaughtException', (error: Error) => {
-
-  logError('Uncaught exception on main process', error)
+function uncaughtException(error: Error) {
 
-  if (sharedProcess) {
-    sharedProcess.console.error('Uncaught exception:')
-    sharedProcess.console.error(error.name)
-    sharedProcess.console.error(error.message)
+  if (hasReportedUncaughtException) {
+    logError(formatError(error))
+    return
   }
 
+  hasReportedUncaughtException = true
+  preventQuit = true
+
+  logError(formatError(error))
```

Comment:
Since we do this in the `hasReportedUncaughtException` branch and here, could we pull this above that to eliminate the duplication?